feat(promql): Add PromQL->Flux transpiler and Flux helper functions

juliusv opened the pull request:
### Done checklist
- [-] docs/SPEC.md updated
- [X] Test cases written

juliusv commented:
As discussed with @nathanielc, this adds the initial implementation of the Flux->PromQL transpiler, including both the transpiler itself and some new Flux helper functions that the transpiler requires to do its job.

juliusv commented:
@nathanielc updated this to use the new `with` operator in `map()` invocations.

nathanielc commented on stdlib/promql/timestamp.go:
```diff
@@ -0,0 +1,146 @@
+		// Get the "_time" (or "_stop") of the current row as a Unix timestamp.
+		for i := 0; i < cr.Len(); i++ {
+			v := execute.ValueForRow(cr, i, timeIdx)
+			ts := float64(v.Time().Time().UnixNano()/1e6) / 1e3
```
This can be simplified to

```
s := float64(v.Time()/1e6) / 1e3
```

The `values.Time` is an `int64` that represents unix nanoseconds.

juliusv replied:
Done, thanks.

wolffcm commented on go.mod:
```diff
@@ -1,41 +1,41 @@
-	github.com/inconshreveable/mousetrap v1.0.0 // indirect
+	github.com/goreleaser/goreleaser v0.97.0
+	github.com/hashicorp/go-immutable-radix v1.1.0 // indirect
+	github.com/influxdata/influxdb v0.0.0-20190603235743-3a07e7f72186
```
It does seem to be only used for testing.  I suppose we could set the dependency to use the tip of `influxdb@flux-master`.  I guess a circular dependency between repos doesn't break the build as long the packages themselves don't depend on each other circularly.

However, adding this dependency seems like it would add complexity.  Right now our dependencies with respect to `influxdb` are very clear, and `flux` is pretty isolated.  The tests run fast and are rarely flaky.

In the short term, we could move just the promql transpiler end-to-end tests into `influxdb`?  Longer term, maybe the PromQL transpiler belongs in `influxdb` as well, similar to the InfluxQL transpiler?

juliusv replied:
Yeah, this cross-dependency annoys me as well, but the test launcher that's in the InfluxDB repo solved too many things to not reuse it :)

> In the short term, we could move just the promql transpiler end-to-end tests into influxdb?

Happy to do that if people agree that's the way to go. Drawback: I don't have many unit tests at the moment, so then we'll depend on end-to-end tests from another repo to make sure that this repo is still working correctly.

The end-to-end tests do take a while to run in any case (due to Flux slowness), so that would make the InfluxDB tests slower to run then.

> Longer term, maybe the PromQL transpiler belongs in influxdb as well, similar to the InfluxQL transpiler?

I'm fine with wherever you all prefer :) I guess the custom Flux functions to make the transpiler work would still have to live in the Flux repo, so then transpiler functionality would be split across repos. But maybe that's ok.

wolffcm replied:
What we used to do in `flux` to do end-to-end testing without depending on `influxdb`, is take any instance of `influxdb.from` and replace it with `csv.from`.  The tests would run against data that we had pregenerated into .csv files.

In the test, before executing the Flux, could we walk the `*ast.File` produced by `Transpile()` and replace each call to `from` with `csv.from`?  Then it wouldn't be necessary to spin up a full stack.  There seems to be just one data set used in the test, so it could be converted to csv and stored in the `testdata` directory.

juliusv replied:
Yeah, that's actually an option I looked at at the beginning before deciding to go for the full InfluxDB test launcher. My worry with using CSV is that I assume that the Flux would execution would get even way slower, which is already so slow (see https://github.com/influxdata/flux/issues/1434) for this tiny dataset. But I may be wrong and can verify this.

Plus, the launcher-based test is more end-to-end, but that can arguably be a separate test in the InfluxDB repo later.

juliusv replied:
Hmm just doing an ad-hoc manual test of `from()` vs. `csv.from()`, CSV seems to add around a second of execution time to each query, which wouldn't really be acceptable IMO, given that I'm running >500 queries already in the test suite?

wolffcm replied:
@juliusv Sorry for the delay in response, it was a long holiday weekend here.

I agree that an additional 500s of testing time is too long.  Can you file an issue detailing the steps to reproduce?

Maybe the best way forward is: 
- Merge this as-is if it passes CI
- Prioritize the bug that is 1s latency being observed using `csv.from`
- Update the PromQL transpiler to use `csv.from` and break the dependency on `influxdb`.

@nathanielc Does this seem reasonable?

juliusv replied:
@wolffcm Happy to file an issue like you described, but I don't know if I would consider that a bug, but expected behavior. If I re-read a CSV file on every query (using `csv.from()`), isn't it supposed to be much slower than if I first ingest all data into a proper DB and theny query it from there?

wolffcm replied:
Hmm, how many rows and tables are in the file?

juliusv replied:
@wolffcm Hmm now I actually noticed that the CSV I ran it against wasn't even complete because I used the web interface to download it from InfluxDB, and I think I got only the response of an earlier query that didn't have all data, because querying for everything yields "Flux response too large". So I'd expect it to be even slower than I thought.

In line-protocol terms, what I'm feeding in is ~450k lines / samples long, which is 60MB of line protocol input data (live-generated from the ~6.6MB "real" Prometheus TSDB that I included in this PR). So that's definitely big to re-read on every query, but overall still tiny in terms of datasets. I could craft a much smaller dataset manually, but I thought it'd be a good start to have something real-ish for a start for end-to-end query comparison tests.

juliusv replied:
(btw. are there better ways to convert line protocol to CSV, or to dump the entire InfluxDB database as CSV?)

wolffcm replied:
Ah, ok, I'm understanding better now.  In Flux all of our end-to-end tests are generally done with toy data sets just large enough to verify correct behavior.  In addition, we embed the actual csv right in the `.flux` script, so there is no disk I/O.  (`csv.from` will accept a `csv` argument.) 

We run about 140 tests in 5 seconds (on my laptop) this way.  This is still kind of slow for a unit test, but we deem it acceptable.

The downside of this approach is that the data set needs to be carefully crafted in order for the tests to have good coverage.  Do you think it would be possible to some up with a small version of this data set (maybe dozens of rows instead of tens of thousands) to exercise the transpiler?

To get raw CSV from a running instance of Influx 2.0, you can use this template:

```sh
curl -v -H "Authorization: Token ${INFLUX_TOKEN}" -H 'Content-Type: application/vnd.flux' -X POST --data-binary "${query}" 'http://localhost:9999/api/v2/query?org=${org}'
```

Just replace `${query}` with your Flux query, and use your own org and token.  You'll probably want to drop the `_start` and `_stop` columns to get whatever data you need since it's the `range` function adds them in, e.g, `from(bucket: "telegraf") |> range(start: -1m) |> drop(columns: ["_start", "_stop"])`.

juliusv replied:
> Do you think it would be possible to some up with a small version of this data set (maybe dozens of rows instead of tens of thousands) to exercise the transpiler?

Yeah, I probably should just work on that now. I wanted to create smaller, case-specific test data anyway, but it'll take me a while. So it was easiest at first to just make Prometheus scrape some data and test against that. It would still be good to keep the larger (yet still relatively small) end-to-end tests around in some form at the end, but maybe those can go into the InfluxDB repo instead.

In any case I'll have to structure my tests a bit differently than the unit tests in https://github.com/influxdata/flux/tree/master/stdlib/testing/testdata, since the Flux queries will be generated by the transpiler and thus can't be stored. So I'm planning on using the strategy of replacing `from()` calls with `csv.from()` calls in the transpiled Flux AST, and maybe also inject the CSV contents directly into the transpiler-generated Flux AST, so we avoid disk IO. (and then I store the same testdata in a Prometheus TSDB, perform equivalent queries, do the necessary output transformations, and compare PromQL vs. transpiled results as I do now). Sound good?

Should this block merging, or is it fine to do later? I'm fine with either. I have to do the occasional master-merge here to resolve go.mod conflicts, but otherwise I don't expect too much work to keep this branch mergeable.

> To get raw CSV from a running instance of Influx 2.0, you can use this template:

Thanks! :)

juliusv replied:
Hmm, I just think that the CSV format is both really hard to write and read and very verbose even for small data... thinking of adopting the Prometheus style of defining test data for this (and then convert to CSV from that), as I'll anyways have to feed it into Prometheus too, e.g.:

https://github.com/prometheus/prometheus/blob/master/promql/testdata/functions.test#L65-L68

juliusv replied:
Though the code for parsing those PromQL testdata definitions is not exposed and quite complex, so maybe a bit much to copy as well. Hrm... still thinking about the best approach.

wolffcm replied:
I agree that csv is painful to work with.  I have never been able to write even simple csv data and have it parse correctly the first time.

I'm happy to defer to whatever @nathanielc wants to do, but my preference would be to avoid creating a dependency on `influxdb` in `flux` master, so my preference would be to block until the testing issue is resolved.

nathanielc replied:
I see a few kinds of tests at play:

1. Tests of the transpilation from PromQL to Flux AST
2. End to End tests of the promql flux helper methods.
3. End to End tests of the transpiler that uses live data against a live prometheus  and a live influxdb.

Do we have any tests that validate that a given PromQL string produces an expected Flux AST? I don't remember if I saw them during review...


The end to end tests for the promql helper functions can use the existing Flux csv end to end tests with small data sets that validate they perform as expected. 



The ability to use live dbs for the end to end tests of the transpiler is valuable as we don't have to write lots of test fixtures and update them. We can easily create a large cross product of tests against many features and parameters and validate they all work. These tests can move to the influxdb repo, so we still get valuable complete coverage.

Those are my thoughts as it will test each piece in isolation and then still have easy to maintain tests that test the entire system. 

Thoughts?

juliusv replied:
> 1. Tests of the transpilation from PromQL to Flux AST
> 2. End to End tests of the promql flux helper methods.
> 3. End to End tests of the transpiler that uses live data against a live prometheus and a live influxdb.
>
> Do we have any tests that validate that a given PromQL string produces an expected Flux AST? I don't remember if I saw them during review...

There's currently two types of tests:

- Unit tests for the Flux helper methods. There are not many unit tests here yet, but they would only cover a tiny bit of the transpiler's functionality anyway (thus medium-useful). Will still add more of those of course.

- End-to-end query comparison tests, which we were talking about here, and which:
  - initialize Prometheus and InfluxDB with the same dataset
  - run a bunch of query variants against both DBs
  - compare & check that results from both systems are equivalent (not requiring any output fixture data, since you can just compare to PromQL, which is the canonical "spec")

Thus I don't have any tests that just check whether a given PromQL string produces an expected Flux AST, as I think the better and less brittle test is just to verify that the actual behavior is identical to a PromQL execution.

So would your suggestion be to move the latter tests to the InfluxDB repo? It's just that they cover most of the functionality... but if we are fine testing that from a different repo for now, that's ok with me too.

juliusv replied:
To elaborate why I think adding PromQL AST -> Flux AST conversion tests with fixed input / output fixtures wouldn't be very useful:

- if underlying Flux behavior changes in a breaking way, we wouldn't notice
- you can't actually realistically read the fixtures as a human and tell whether they are "right"
- the fixtures would change with any little implementation change, even if behavior stays the same

nathanielc replied:
Ok, that works for me. 

So to merge lets do two things:

1. Move the `End to End tests of the transpiler that uses live data against a live prometheus and a live influxdb.` to the influxdb repo
2. Add pure Flux e2e test  for each of the promql  helper methods. These tests can assume the data is correctly structured.

juliusv replied:
@nathanielc Sounds good, will do!

juliusv replied:
@nathanielc I added Flux end-to-end tests for the helper functions in this commit: https://github.com/influxdata/flux/pull/1419/commits/222c0742333b33607b5b284dddf031b5bfc73d4f

While at it I simplified the input/output column assumptions of those functions a bit.

juliusv replied:
Ah, the way that unit tests (not e2e Flux tests) work changed in master, which broke my one `promql.histogramQuantile()` unit test I already had. Will fix that.

juliusv replied:
Alright, fixed that as well.

nathanielc commented on compile.go:
```diff
@@ -87,6 +87,7 @@ var (
 		"universe",
 		"influxdata/influxdb",
 		"math",
+		"promql",
```
I agree, we do not want it part of the prelude. Uses should explicitly import `promql` if they need it.

juliusv replied:
Yeah, I think it just needs to be here at the moment so that `promql` functions can be used as lambdas.

juliusv commented on promql/binary_expr.go:
```diff
@@ -0,0 +1,387 @@
+			), nil
+		}
+
+		panic(fmt.Errorf("invalid scalar-vector binary op %q", b.Op))
```
Yes, there's a couple of cases where I rely on the PromQL parser to already verify things and panic if that assumption is violated (like here). I'll convert those panics to errors then.

Note that there'll still be some places where I make those assumptions implicitly without an explicit check&panic for everything.

juliusv replied:
(Hmm, keeping the panics in the label matchers transpilation for now as it would otherwise blow up the code needlessly, since there's no errors expected in that code path so far. But converted the other instances).

juliusv commented on promql/transpile.go:
```diff
@@ -0,0 +1,274 @@
+	default:
+		lastCall := calls[len(calls)-1]
+		// Skip nil call arguments.
+		if lastCall == nil {
```
When building pipelines I sometimes have optional parts that get set to some value if a certain condition is true, and otherwise remain `nil`. Then I don't have to have a lot of conditional concatenation logic at the end when I build a pipeline. I added a comment about that here.

juliusv commented on promql/transpile.go:
```diff
@@ -0,0 +1,274 @@
+// Transpile converts a PromQL expression with the time ranges set in the transpiler
+// into a Flux file. The resulting Flux file can be executed and the result needs to
+// be transformed using FluxResultToPromQLValue() to get a result value that is fully
+// equivalent to the result of a native PromQL execution.
```
Thanks :)

juliusv commented on promql/unary_expr.go:
```diff
@@ -0,0 +1,41 @@
+		}
+
+		// Multiply float expression by -1.
+		return &ast.BinaryExpression{
```
Thanks for pointing that out! I changed the non-table float minus operator case to using `ast.UnaryExpression`.

juliusv commented on stdlib/promql/changes.go:
```diff
@@ -0,0 +1,177 @@
+	"github.com/influxdata/flux/semantic"
+)
+
+const ChangesKind = "changes"
```
Ended up adding them to `promql.flux`.

juliusv commented on stdlib/promql/date_functions.go:
```diff
@@ -0,0 +1,73 @@
+				return nil, fmt.Errorf("cannot convert argument of type %v to float", v.Type().Nature())
+			}
+
+			t := time.Unix(0, int64(v.Float()*1e9)).UTC()
```
Done.

juliusv commented on stdlib/promql/histogram_quantile.go:
```diff
@@ -0,0 +1,396 @@
+}
+
+// Taken from Prometheus.
+// TODO: Include some kind of copyright notice?
```
In general the rules are pretty clear that I'd have to retain the original license statement, it just gets complicated-ish when you take little bits of code here and there (mixed with new code) and you still try to give proper attribution everywhere :) Probably it's fine to just include a license comment at the top and state that some functionality was adapted from Prometheus?

juliusv commented:
Thanks for the review @nathanielc and @wolffcm, very helpful and great to hear that it's mainly in line with expectations :)

I think I addressed all the comments that were made so far.

juliusv commented:
I have now also removed the transpiler end-to-end tests from this repo, thus getting rid of the dependency on the InfluxDB repo. I also added the missing copyright notice in `histogram_quantile.go` for functions that were verbatim copied from Prometheus.

Should hopefully be all fine to merge now.

nathanielc commented:
@juliusv This looks good. Did you create an PR against influxdb repo with the tests? 

I have asked @wolffcm to give this another review and then I think we can merge.

juliusv commented:
@nathanielc 

> Did you create an PR against influxdb repo with the tests?

It's ready and just waiting to be filed once this is merged, since it depends on it. But you can see the e2e tests in this branch (modulo the dep updates after this merge): https://github.com/juliusv/influxdb/commits/promql-transpiler-e2e-tests

juliusv commented:
FYI, somehow I hadn't pushed half of the `label_replace()` functionality, I added that now, along with Flux e2e tests for it.

wolffcm commented:
Hey @juliusv This looks good to me.

I cloned your branch into the Flux repo and it ran in CI.
Currently it's failing `staticcheck` because of unused functions:
```
promql/binary_expr.go:182:6: func stripSuffixFn is unused (U1000)
promql/binary_expr.go:199:5: var matchRHSSuffixFn is unused (U1000)
```
Can you fix this and make sure that `make staticcheck` passes?  All the other parts of CI looked good.

Other than that this looks good to go.
